### ISSUE TITLE
Assessing Generalization in Deep Reinforcement Learning  The Berkeley Artificial Intelligence Research Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5523]
Assessing Generalization in Deep Reinforcement Learning &ndash; The Berkeley Artificial Intelligence Research Blog<br>
<br>
https://ift.tt/2UKHOM0<br>
<br>
Daniel Seita<br>
<br>
<div>
<article>
<p><img width="600" src="http://bair.berkeley.edu/static/blog/rl_gen/6_panel_1080_24.gif"></p>
<h2>TL;DR</h2>
<p>We present a benchmark for studying generalization in deep reinforcement learning (RL). Systematic empirical evaluation shows that vanilla deep RL algorithms generalize better than specialized deep RL algorithms designed specifically for generalization. In other words, simply training on varied environments is so far the most effective strategy for generalization. The code can be found at <a href="https://github.com/sunblaze-ucb/rl-generalization">https://github.com/sunblaze-ucb/rl-generalization</a> and the full paper is at <a href="https://arxiv.org/abs/1810.12282">https://arxiv.org/abs/1810.12282</a>.</p>
<h2>Motivation</h2>
<p>The ability to adapt to new, unseen situations and environments is a hallmark of human intelligence. For example, drivers are able to quickly adjust to changes in the road surface and weather conditions, new traffic rules, and idiosyncrasies of particular vehicles. Humans are able to adapt their driving to new scenarios because they have a representation of the world that generalizes beyond the city in which they normally drive.</p>
<p>Deep RL has emerged as an important family of techniques that may support the development of intelligent systems. Recent advances driven by deep RL include achieving human-level performance on complex games such as <a href="https://deepmind.com/research/publications/human-level-control-through-deep-reinforcement-learning/">Atari</a>, <a href="https://deepmind.com/research/alphago/">Go</a>, and <a href="https://deepmind.com/blog/alphastar-mastering-real-time-strategy-game-starcraft-ii/">Starcraft</a>.</p>
<p>As excitement surrounding deep RL continues to grow, so is awareness of limitations and pitfalls such as overfitting: the tendency of policies trained via RL to specialize to their training domain and break down when deployed in different circumstances. <a href="https://arxiv.org/abs/1709.06009">Several</a> <a href="https://arxiv.org/abs/1812.02850">recent</a> papers used Atari games to show that deep RL is susceptible to dramatically overfit to the idiosyncrasies of training environments. This problem is not limited to Atari: severe overfitting has also been demonstrated in <a href="https://arxiv.org/abs/1806.07937">other domains</a>. A major contributor to this problem is the fact that deep RL agents are commonly trained and tested in the same environment and are thus not encouraged to learn representations that generalize to previously unseen circumstances. As deep RL enters the mainstream, it is clear that the brittleness of such systems to minor changes in the environment can have serious implications for real-world systems, and this problem must be addressed before applying deep RL to problems like autonomous driving, where errors may be catastrophic.</p>
<p>Generalization in deep RL has been recognized as an important problem and is under active investigation. In the past year, OpenAI has released two benchmarks for generalization in deep RL: the <a href="https://openai.com/blog/retro-contest/">Retro contest</a>, which tests if a game-playing AI generalizes to previously unseen levels of the same game (Sonic The Hedgehog), and <a href="https://github.com/openai/coinrun">CoinRun</a>, a new game environment that tests agents&rsquo; ability to generalize to new levels created using a procedural generator. <a href="https://arxiv.org/abs/1806.10729">Justesen et al. (2018)</a> used procedural generation of video game levels during training to improve generalization to human-designed levels at test time. Earlier, DeepMind released a suite of <a href="https://deepmind.com/research/publications/ai-safety-gridworlds/">&ldquo;AI safety&rdquo; gridworlds</a> designed to test the susceptibility of RL agents to scenarios that can trigger unsafe behavior (spoiler: state-of-the-art deep RL algorithms are highly fallible and exhibit unsafe behavior). Despite these and other efforts, there is a lack of consistency in the experimental methodology adopted in the literature. Each project uses a different set of evaluation metrics, environments, and environment variations over which agents are expected to generalize. The merits of different algorithms thus remain difficult to compare.</p>
<p>Our contribution is a framework for investigating generalization in deep RL with a clearly defined set of environments (based on those already common in deep RL research), environment variations, and evaluation metrics. We consider variations in environment dynamics, such as when the road becomes slippery due to rain, rather than variations in the environment goal, such as driving to a new location in your hometown.</p>
<p>As a baseline, we also conduct a systematic empirical study of the generalization performance of deep RL algorithms, including recent specialized techniques for tackling generalization in deep RL. We differentiate between interpolation to environments similar to those seen during training and extrapolation to environments that are beyond the ranges observed during training. Good extrapolation performance is crucial to the safe deployment of RL agents to the real world.</p>
<h2>Environments</h2>
<p>We build on four classic control environments from <a href="https://gym.openai.com/">OpenAI Gym</a> and two locomotion environments from <a href="https://blog.openai.com/roboschool/">Roboschool</a> that have been used in prior work on generalization in deep RL. Specifically, we introduce new versions of Acrobot, CartPole, MountainCar, Pendulum, HalfCheetah, and Hopper.</p>
<p>To test agents&rsquo; ability to generalize to changes in environment dynamics, the environments are modified so that we can vary certain parameters that directly affect the dynamics. For example, for the two locomotion environments (HalfCheetah and Hopper), we vary the parameters for robot power, torso density, and joint friction.</p>
<p><img width="600" src="http://bair.berkeley.edu/static/blog/rl_gen/cheetah%20friction.mp4"><br>
<i>HalfCheetah under varying joint frictions.</i></p>
<p>Each environment has three versions:</p>
<ul>
<li>
<p>Default (D), where the parameters are fixed at the default values used in Gym or Roboschool.</p>
</li>
<li>
<p>Random (R), where the parameters are uniformly sampled from ranges surrounding the values in D. Practically, this represents the distribution of environments from which it is possible to obtain training data.</p>
</li>
<li>
<p>Extreme (E), where the parameters are uniformly sampled outside the ranges used in R. The values are more extreme than those in R. The Extreme mode represents edge cases and other environments for which it is difficult to obtain training data.</p>
</li>
</ul>
<p><img width="600" src="http://bair.berkeley.edu/static/blog/rl_gen/DRE_figure.png"><br>
<i>In CartPole, we vary two parameters, the length and the mass of the pole; the left panel shows a schematic of the parameter ranges in each version. In the right panel we illustrate the environment when the length falls in the ranges of the three versions (E in <b>red</b>, R in <b>green</b>, and D as a fixed point in R). The E range excludes the R (and D) range.</i></p>
<h2>Evaluation metrics</h2>
<p>In the spirit of RL as <a href="http://incompleteideas.net/book/the-book.html">goal-seeking</a>, we measure the probability of achieving a predefined goal instead of a scalar reward. Doing so has the advantages of being invariant to the environment parameters and unaffected by reward shaping. Though reward shaping can be useful in getting agents to learn desired behaviors, it creates problems when using reward to compare the merits of various algorithms, since the reward equations are often not easily understood and hard to compare, as well as prone to change across environment versions. For example, the equation used to calculate reward in Roboschool&rsquo;s HalfCheetah is an obtuse combination of body positioning, electricity, torque, collisions, etc. A poorly constructed reward can also lead to trained agents with <a href="https://openai.com/blog/faulty-reward-functions/">undesirable behavior</a>. Instead of directly using reward as a success metric, we chose goals with a real-world meaning; for example, on HalfCheetah the goal is to walk 20 meters, and the reported success is the percent of attempts in which the agent achieves this goal.</p>
<p>We design a set of three metrics for generalization performance.</p>
<ol>
<li>
<p>Default: success rate on D when trained on D. This is the classic RL setting and thus serves as a baseline for the other two metrics.</p>
</li>
<li>
<p>Interpolation: success rate on R when trained on R. This measures the performance of an agent on environments similar to, but not exactly the same as, those seen during training.</p>
</li>
<li>
<p>Extrapolation: geometric mean of the success rates on E when trained on R, on R when trained on D, and on E when trained on D. This measures the performance of an agent on environments different from those seen during training.</p>
</li>
</ol>
<h2>Results</h2>
<p>We first evaluate two popular deep RL algorithms, one from the policy gradient family, <a href="https://openai.com/blog/openai-baselines-ppo/">PPO</a>, and one from the actor-critic family, <a href="https://openai.com/blog/baselines-acktr-a2c/">A2C</a>. We then evaluate two specialized methods for training agents that generalize: <a href="https://arxiv.org/abs/1610.01283">EPOpt</a> and <a href="https://arxiv.org/abs/1611.02779">RL$^2$</a>. EPOpt trains an agent to be robust to environment variations by maximizing a risk-sensitive reward while RL$^2$ aims to learn a policy that can adapt to the environment at hand using the observed trajectory. As general-purpose techniques, EPOpt and RL$^2$ are each combined with the two vanilla deep RL algorithms.</p>
<p>The following chart shows the three generalization metrics discussed above, averaged over the six environments and then summarized over five runs of the experiments. FF refers to a feed-forward network architecture for the policy/value function and RC refers to a recurrent one. For visual clarity, we only show a subset of the algorithms here (see the paper for the complete table of results).</p>
<p><img width="500" src="http://bair.berkeley.edu/static/blog/rl_gen/benchmark_results_vanilla.png"></p>
<p>Aside from PPO with the recurrent network architecture, A2C and PPO were able to interpolate fairly well but had limited extrapolation success. In other words, simply training on a random environment, without adding any specialized mechanism, results in agents that can generalize to similar environments.</p>
<p><img width="500" src="http://bair.berkeley.edu/static/blog/rl_gen/benchmark_results_generalize.png"></p>
<p>EPOpt provided improved interpolation and extrapolation performance, but only when combined with PPO and a feed-forward architecture. In fact, the improvement was seen in the environments with continuous action spaces (Pendulum, HalfCheetah, and Hopper), indicating that this may be important to the success of EPOpt. In other cases, EPOpt does not perform as well as the vanilla deep RL algorithms.</p>
<p>RL$^2$ proved difficult to train, leading to poor generalization performance. In most cases, a working policy was not found even for the fixed environment.</p>
<h3>Case study: HalfCheetah</h3>
<p>Differences in gait are readily observable in HalfCheetah, making it a good candidate for a case study. We look at EPOpt-PPO with the feed-forward architecture, the best-performing algorithm on this environment, and compare it to PPO with the feed-forward architecture.</p>
<p><img width="600" src="http://bair.berkeley.edu/static/blog/rl_gen/4_panel_5s_1080_24.gif"></p>
<p>When trained on the fixed environment, EPOpt-PPO learns an agent that, compared to PPO, takes smaller steps. Because EPOpt is designed to be robust, this makes intuitive sense; when faced with an unfamiliar (possibly dangerous) terrain, humans do very much the same in order to avoid falling. When trained on the random environment, compared to the fixed environment, both PPO and EPOpt-PPO learn slower gaits with the head leaning forward and downwards. We hypothesize that such a body position helps the robot find balance across a wider range of environmental parameters.</p>
<p>The <a href="https://openai.com/blog/first-retro-contest-retrospective/">results of the OpenAI Retro contest</a>, the <a href="https://arxiv.org/abs/1812.02341">paper accompanying CoinRun</a>, and <a href="https://arxiv.org/abs/1806.10729">Justesen et al. (2018)</a> also suggest that training with environment stochasticity (i.e. on a distribution of environments) is an effective strategy for generalization. In a similar vein, <a href="https://arxiv.org/abs/1806.07937">Zhang et al. (2018)</a> show that in a fixed environment, using multiple random seeds to generate trajectories during training improves generalization to possible changes at test time.</p>
<h2>Moving forward</h2>
<p>The results of our empirical evaluation have shown that it is crucial to perform a thorough study of proposed algorithms on a variety of tasks. Despite EPOpt being an intuitively general-purpose approach, it only showed promising results with PPO and a feed-forward policy network. The difficulty in training RL$^2$ suggests that a reward signal alone may not be sufficient for an LSTM to learn something useful about the environment at hand from the current trajectory, and so a more structured architecture or a model-based approach may be needed for robust automatic test-time adaptation.</p>
</article>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via The Berkeley Artificial Intelligence Research Blog https://ift.tt/2EPVVfk<br>
<br>
September 27, 2019 at 11:53AM
